src/network/networkd-ndisc.c

```diff
@@ -1175,15 +1226,33 @@ static int ndisc_router_handler(Link *link, sd_ndisc_router *rt) {
         if (r < 0)
                 return r;
 
-        r = ndisc_drop_outdated(link, timestamp_usec);
+        r = sd_ndisc_router_get_lifetime(rt, &router_lifetime_sec);
         if (r < 0)
-                return r;
+                return log_link_warning_errno(link, r, "Failed to get lifetime of RA message: %m");
 
-        r = ndisc_start_dhcp6_client(link, rt);
+        r = ndisc_drop_outdated(link, timestamp_usec);
         if (r < 0)
                 return r;
```

Comment:
Please do not mix the order.
1. check router address and filter the message if necessary,
2. check lifetime and drop previous assignments if lifetime is zero,
3. check timestamp and drop outdated assignments if necessary.

```c
r = sd_ndisc_router_get_lifetime(rt, &router_lifetime_sec);
if (r < 0)
        return log_link_warning_errno(link, r, ...);

if (router_lifetime_sec == 0) {
        r = ndisc_process_zero_lifetime(...);
        if (r < 0)
                log_link_warning_errno(...);
        return 0;
}

r = sd_ndisc_router_get_timestamp(rt, CLOCK_BOOTTIME, &timestamp_usec);
if (r == -ENODATA)
        ...
if (r < 0)
        ...

r = ndisc_drop_outdated(link, timestamp_usec);
if (r < 0)
        return r;

r = ndisc_start_dhcp6_client(...);
if (r < 0)
        return r;
```